Added documentation for typegen & the VS Code extension (#2963)
statelyai/xstate: docs/guides/typescript.md

> This feature is in beta! See the section on known limitations below to see what we're actively looking to improve.

Here's how you can get started:

1. Download and install the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode).
2. Open a new file and create a new machine, passing the schema attributes:

```ts
import { createMachine } from 'xstate';

const machine = createMachine({
  schema: {
    context: {} as { value: string },
    events: {} as { type: 'FOO'; value: string } | { type: 'BAR' }
  },
  initial: 'a',
  states: {
    a: {
      on: {
        FOO: {
          actions: 'consoleLogValue',
          target: 'b'
        }
      }
    },
    b: {
      entry: 'consoleLogValueAgain'
    }
  }
});
```

3. Add `tsTypes: {}` to the machine and save the file:

```ts
const machine = createMachine({
  tsTypes: {},
  schema: {
    context: {} as { value: string },
    events: {} as { type: 'FOO'; value: string } | { type: 'BAR' }
  },
  initial: 'a',
  states: {
    /* ... */
  }
});
```

4. The extension should automatically add a generic to the machine:

```ts
const machine = createMachine({

>   tsTypes: {} as import('./filename.typegen').Typegen[0]

Reviewer: ```suggestion
  tsTypes: {} as import('./filename.typegen').Typegen0
```

const lightMachine = createMachine({
  schema: {
    // The context (extended state) of the machine
    context: t<{ elapsed: number }>(),
    // The events this machine handles
    events: t<
      | { type: 'TIMER' }
      | { type: 'POWER_OUTAGE' }
      | { type: 'PED_COUNTDOWN'; duration: number }
    >()
  }
  /* Other config... */
});
```

Providing the context and events to the `schema` attribute gives many advantages:

- The context type/interface (`TContext`) is passed on to actions, guards, services and more. It is also passed to deeply nested states.
- The event type (`TEvent`) ensures that only specified events (and built-in XState-specific ones) are used in transition configs. The provided event object shapes are also passed on to actions, guards, and services.
- Events which you send to the machine will be strongly typed, offering you much more confidence in the payload shapes you'll be receiving.

## Typegen with the VSCode Extension <Badge text="4.29+" />

Using our [VS Code extension](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode), you can automatically generate intelligent typings for XState.

> > This feature is in beta! See the section on known limitations below to see what we're actively looking to improve.

Reviewer: ```suggestion
::: warning
This feature is in beta! See the section on known limitations below to see what we're actively looking to improve.
```

Author: Changed it to this:

![image](https://user-images.githubusercontent.com/28293365/151373174-0bd06775-111e-4055-a77d-de1057121bfa.png)

As XState is written in [TypeScript](https://www.typescriptlang.org/), strongly typing your statecharts is useful and encouraged. Consider this light machine example:

```typescript
const lightMachine = createMachine({
  schema: {
    // The context (extended state) of the machine
    context: t<{ elapsed: number }>(),
    // The events this machine handles
    events: {} as
      | { type: 'TIMER' }
      | { type: 'POWER_OUTAGE' }
      | { type: 'PED_COUNTDOWN'; duration: number }

>   },

Reviewer: I would probably stick with a single way of providing those types throughout the documentation. Using stuff consistently raises less questions from the readers

});
```

Providing the context and events to the `schema` attribute gives many advantages:

- The context type/interface (`TContext`) is passed on to actions, guards, services and more. It is also passed to deeply nested states.
- The event type (`TEvent`) ensures that only specified events (and built-in XState-specific ones) are used in transition configs. The provided event object shapes are also passed on to actions, guards, and services.
- Events which you send to the machine will be strongly typed, offering you much more confidence in the payload shapes you'll be receiving.

## The VSCode Extension <Badge text="4.29+" />

> Using our [VS Code extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode), you can automatically generate intelligent typings for XState.

Reviewer: ```suggestion
Using our [VS Code extension](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode), you can automatically generate intelligent typings for XState.
```

Here's how you can get started:

> 1. Download and install the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode).

Reviewer: ```suggestion
1. Download and install the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode).
```

2. Open a new file and create a new machine, passing the schema attributes:

```ts
import { createMachine, t } from 'xstate';

const machine = createMachine({
  schema: {
    context: t<{ value: string }>(),
    events: t<{ type: 'FOO'; value: string } | { type: 'BAR' }>()
  },
  initial: 'a',
  states: {
    a: {
      on: {
        FOO: {
          actions: 'consoleLogValue',
          target: 'b'
        }
      }
    },
    b: {
      entry: 'consoleLogValueAgain'
    }
  }
});
```

> 3. Add `tsTypes: true` to the machine and save the file:

Reviewer: Perhaps this should use something that is more compatible with the injected replacement
```suggestion
3. Add `tsTypes: {}` to the machine and save the file:
```
```suggestion
3. Add `tsTypes: {}` to the machine and save the file:
```

```ts
const machine = createMachine({

>   tsTypes: true,

Reviewer 2: Above it says "Add `tsTypes: {}` ...". What does `true` mean here?

Author: True, this shouldn't be here! An old version snuck in.

});
```

5. Add a second parameter into the `createMachine` call - this is where you implement the actions, services, guards and delays for the machine.

```ts
const machine = createMachine(
  {
    // Config here
  },
  {
    actions: {
      consoleLogValue: (context, event) => {
        // Wow! event is typed to { type: 'FOO' }
        console.log(event.value);
      },
      consoleLogValueAgain: (context, event) => {
        // Wow! event is typed to { type: 'FOO' }
        console.log(event.value);
      }
    }
  }
);
```

You'll notice that the events in the options are _strongly typed to the events that cause the action to be triggered_. This is true for actions, guards, services and delays.

You'll also notice that `state.matches`, `tags` and other parts of the machine are now type-safe.

### How to get the most out of the VS Code extension

#### Use named actions/guards/services

Our recommendation with this approach is to mostly use named actions/guards/services, not inline ones.

This is optimal:

```ts
createMachine(
  {
    entry: ['sayHello']
  },
  {
    actions: {
      sayHello: () => {
        console.log('Hello!');
      }
    }
  }
);
```

> This is useful, but less optimal:

Reviewer: this will prevent loading it in the editor, right? so in a way - it's not that useful 🤣

Author: Iiiish, you'll still be able to load the inspector, and generate typegen from it.

Reviewer: should we even allow generating typegen with generics? what would happen in such a situation? would it be pushed to the last place of the generics list? would it be inserted at the position of `tsTypes` property?

Author: It gets inserted at the position - then you get a red line at the generics saying 'refactor this to schema'

```ts
createMachine({
  entry: [
    () => {
      console.log('Hello!');
    }
  ]
});
```

Named actions/services/guards allow for:

- Better visualisation, because the names appear in the statechart
- Easier-to-understand code
- Overrides in `useMachine`, or `machine.withConfig`

> #### Typing onDone/onError functions

Author: ```suggestion
#### Typing promise services
```

> You can use the generated types to specify the result of `onDone` services, by using the `services` schema property:

Author: ```suggestion
You can use the generated types to specify the return type of promise-based services, by using the `services` schema property:
```

});
```

Providing the context and events to the `schema` attribute gives many advantages:

- The context type/interface (`TContext`) is passed on to actions, guards, services and more. It is also passed to deeply nested states.
- The event type (`TEvent`) ensures that only specified events (and built-in XState-specific ones) are used in transition configs. The provided event object shapes are also passed on to actions, guards, and services.
- Events which you send to the machine will be strongly typed, offering you much more confidence in the payload shapes you'll be receiving.

## The VSCode Extension <Badge text="4.29+" />

> Using our [VSCode extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode), you can automatically generate intelligent typings for XState.

Reviewer 3: ```suggestion
Using our [VS Code extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode), you can automatically generate intelligent typings for XState.
```

Here's how you can get started:

> 1. Download and install the [VSCode extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode).

Reviewer 3: ```suggestion
1. Download and install the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode).
```

2. Open a new file and create a new machine, passing the schema attributes:

```ts

> const machine = createMachine({

Reviewer 3: ```suggestion
import { createMachine, t } from 'xstate';

const machine = createMachine({
```

  schema: {
    context: t<{ value: string }>(),
    events: t<{ type: 'FOO'; value: string } | { type: 'BAR' }>()
  },
  initial: 'a',
  states: {
    a: {
      on: {
        FOO: {
          actions: 'consoleLogValue',
          target: 'b'
        }
      }
    },
    b: {
      entry: 'consoleLogValueAgain'
    }
  }
});
```

3. Add `tsTypes: true` to the machine and save the file:

```ts
const machine = createMachine({
  tsTypes: true,
  schema: {
    context: t<{ value: string }>(),
    events: t<{ type: 'FOO'; value: string } | { type: 'BAR' }>()
  },
  initial: 'a',

>   states: {
>     a: {
>       on: {
>         FOO: {
>           actions: 'consoleLogValue',
>           target: 'b'
>         }
>       }
>     },
>     b: {
>       entry: 'consoleLogValueAgain'
>     }
>   }

Reviewer 3: Just for brevity/easier reading:

```suggestion
  states: {/* ... */}
```

});
```

4. The extension should automatically add a generic to the machine:

```ts
const machine = createMachine({
  tsTypes: {} as import('./filename.typegen').Typegen[0],
  schema: {
    context: t<{ value: string }>(),
    events: t<{ type: 'FOO'; value: string } | { type: 'BAR' }>()
  },
  initial: 'a',

>   states: {
>     a: {
>       on: {
>         FOO: {
>           actions: 'consoleLogValue',
>           target: 'b'
>         }
>       }
>     },
>     b: {
>       entry: 'consoleLogValueAgain'
>     }
>   }

Reviewer 3: ```suggestion
  states: {/* ... */}
```

});
```

5. Add a second parameter into the `createMachine` call - this is where you implement the actions, services, guards and delays for the machine.

```ts
const machine = createMachine(
  {
    // Config here
  },
  {
    actions: {
      consoleLogValue: (context, event) => {
        // Wow! event is typed to { type: 'FOO' }
        console.log(event.value);
      },
      consoleLogValueAgain: (context, event) => {
        // Wow! event is typed to { type: 'FOO' }
        console.log(event.value);
      }
    }
  }
);
```

You'll notice that the events in the options are _strongly typed to the events that cause the action to be triggered_. This is true for actions, guards, services and delays.

You'll also notice that `state.matches`, `tags` and other parts of the machine are now type-safe.

> ### How to get the most out of the VSCode extension

Reviewer 3: ```suggestion
### How to get the most out of the VS Code extension
```

      type: 'EVENT_WITHOUT_FLAG';
    };

createMachine(
  {
    schema: {

>       context: t<Context>(),

Reviewer: just an extra argument for `{} as ...` - if somebody lands on a doc site which doesnt mention `import { t } from 'xstate'` they might not be able to easily find our what that `t` is and how to get it. Using cmd/ctrl+F: "t" won't help them much either 😉

Author: @davidkpiano, what do you think of `t()`?
